### DISCUSSION TITLE
Data transformation/normalisation for heatmaps

### DISCUSSION THREAD

**Discussion by Jengel44** [ImmuneDynamics/Spectre#132] [Q&A] [answered]
Hi. 
I'm wondering about how data is transformed/normalised for the heatmap plotting? I have asinh_transformed MFI values ranging from -8 to 10 and when plotted on the heatmap I get a scale from 0-1. Is the mean MFI first calculated in the data group and then normalised? And how is it normalised?
Thanks so much!

**Comment by tomashhurst** [comment, marked as answer]
Hi @Jengel44 , 

Great question! If I have understood your question correctly, you are talking about plotting aggregate expression data on a heatmap -- i.e. the median expression of each marker on each cluster/population. 

From the last bit of this section: https://immunedynamics.io/spectre/simple-discovery/#5_Clustering_and_DR

The `do.aggregate` function does the 'aggregation', so the resulting table is rows = clusters, columns = markers, and the values are the median expression of each marker on each cluster.

```
exp <- do.aggregate(cell.dat, cellular.cols, by = "FlowSOM_metacluster")
```

Then the `make.pheatmap` function has an argument which 'normalises' the values in each column (marker) between 0 and 1. This is more accurately referred to as rescaling. 

```
make.pheatmap(exp, "FlowSOM_metacluster", cellular.cols)
```

This means they all have the same colour range for what is effectively minimum to maximum for each marker. So if you had a range of -8 to 10, this would then be 'rescaled' to a range of 0 and 1.

```
normalise | DEFAULT = TRUE. Logical. Only applies to standard heatmaps (i.e. when is.fold = FALSE). TRUE to normalise each column between 0 and 1, FALSE to plot the raw values.
```

If you wanted to turn this feature OFF:

```
make.pheatmap(exp, "FlowSOM_metacluster", cellular.cols, normalise = FALSE)
```

**Reply by Jengel44** [reply]
Ok thanks so much for clarifying this for me. Great explanation for an R/spectre newbie!
